File: openpifpaf/decoder/utils/caf_scored.py
import logging
import time
from typing import List

import numpy as np

# pylint: disable=import-error
from ...functional import scalar_values
from ... import headmeta

LOG = logging.getLogger(__name__)


class CafScored:
    default_score_th = 0.1
    ablation_no_rescore = False

    def __init__(self, cifhr, *, score_th=None, cif_floor=0.1):
        self.cifhr = cifhr
        self.score_th = score_th or self.default_score_th
        self.cif_floor = cif_floor

        self.forward = None
        self.backward = None

    def directed(self, caf_i, forward):
        if forward:
            return self.forward[caf_i], self.backward[caf_i]

        return self.backward[caf_i], self.forward[caf_i]

    def rescore(self, nine, joint_t):
        if self.cif_floor < 1.0 and joint_t < len(self.cifhr) and not self.ablation_no_rescore:
            cifhr_t = scalar_values(self.cifhr[joint_t], nine[3], nine[4], default=0.0)
            nine[0] = nine[0] * (self.cif_floor + (1.0 - self.cif_floor) * cifhr_t)
        return nine[:, nine[0] > self.score_th]

    def fill_single(self, all_fields, meta: headmeta.Caf):
        start = time.perf_counter()
        caf = all_fields[meta.head_index]

        if self.forward is None:
            self.forward = [np.empty((9, 0), dtype=caf.dtype) for _ in caf]
            self.backward = [np.empty((9, 0), dtype=caf.dtype) for _ in caf]

        for caf_i, nine in enumerate(caf):
            assert nine.shape[0] == 9

            mask = nine[0] > self.score_th
            if not np.any(mask):
                continue
            nine = nine[:, mask]

            if meta.decoder_min_distance:
                dist = np.linalg.norm(nine[1:3] - nine[5:7], axis=0)
                mask_dist = dist > meta.decoder_min_distance / meta.stride
                nine = nine[:, mask_dist]

            if meta.decoder_max_distance:
                dist = np.linalg.norm(nine[1:3] - nine[5:7], axis=0)
                mask_dist = dist < meta.decoder_max_distance / meta.stride
                nine = nine[:, mask_dist]

            nine[(1, 2, 3, 4, 5, 6, 7, 8), :] *= meta.stride

            nine_b = np.copy(nine[(0, 3, 4, 1, 2, 6, 5, 8, 7), :])
            nine_b = self.rescore(nine_b, meta.skeleton[caf_i][0] - 1)
            self.backward[caf_i] = np.concatenate((self.backward[caf_i], nine_b), axis=1)

            nine_f = np.copy(nine)
            nine_f = self.rescore(nine_f, meta.skeleton[caf_i][1] - 1)
            self.forward[caf_i] = np.concatenate((self.forward[caf_i], nine_f), axis=1)

        LOG.debug('scored caf (%d, %d) in %.3fs',
                  sum(f.shape[1] for f in self.forward),
                  sum(b.shape[1] for b in self.backward),
                  time.perf_counter() - start)
        return self

    def fill(self, all_fields, metas: List[headmeta.Caf]):
        for meta in metas:
            self.fill_single(all_fields, meta)

        return self

File: cpp/examples/cifcafdecoder_v0.12.13_wholebody/src/decoder/caf_scored.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <cassert>

#include "../../include/decoder/caf_scored.hpp"
#include "../../include/decoder/functional.hpp"
#include "../../include/utils/numpy_utils.hpp"


CafScored::CafScored() {}

CafScored::CafScored(
    blobnd<float> &cifhr,
    unsigned H, unsigned W,
    const CafMetas &meta
) {
    this->cifhr = cifhr;
    this->meta = meta;
    this->H = H;
    this->W = W;
}

pair<Vector2D, Vector2D> CafScored::directed(int caf_i, bool forward) {
    if(forward) {
        return make_pair(this->forward[caf_i], this->backward[caf_i]);
    }

    return make_pair(this->backward[caf_i], this->forward[caf_i]);
}

Vector2D CafScored::rescore(Vector2D &nine, int joint_t) {
    if(this->cif_floor < 1.0 && joint_t < this->cifhr.shape[0]) {
        vector<float> cifhr_t = scalar_values(this->cifhr, joint_t, nine[3], nine[4], 0.0);
        float cfm = 1.0 - this->cif_floor;
        for(unsigned i = 0; i < cifhr_t.size(); i++) {
            cifhr_t[i] *= cfm;
        }
        for(unsigned i = 0; i < cifhr_t.size(); i++) {
            cifhr_t[i] += this->cif_floor;
        }
        for(unsigned i = 0; i < nine[0].size(); i++) {
            cifhr_t[i] *= nine[0][i];
        }

        nine[0] = cifhr_t;
    }

    vector<bool> mask_1d = mask1d(nine[0], this->default_score_th);
    Vector2D nine_filtered = filter_vector2D(nine, mask_1d);
    return nine_filtered;
}

void CafScored::fill_single(const Fields &all_fields) {
    float *caf = all_fields[this->meta.head_index];
    blobnd<float> caf_(caf, {this->caf_d1, this->caf_d2, this->H, this->W});

    int d1 = caf_.shape[0];
    for(int i = 0; i < d1; i++) {
        blobnd<bool> mask = mask2d(caf_, i, 0, this->default_score_th);

        bool any = false;
        for(int j = 0; j < mask.shape[0]; j++) {
            for(int k = 0; k < mask.shape[1]; k++)
                if(mask(j, k)) {
                    any = true;
                    break;
                }
        }

        if(!any) {
            vector<vector<float>> empty(9, vector<float>(0));
            this->backward.push_back(empty);
            this->forward.push_back(empty);
            continue;
        }

        Vector2D nine_filtered = filter_vector3D(caf_, i, 0, mask);

        if(this->meta.decoder_min_distance) {
            const Vector2D &nine1 = {nine_filtered[1], nine_filtered[2]};
            const Vector2D &nine2 = {nine_filtered[3], nine_filtered[4]};
            Vector2D sub = subtract2d(nine1, nine2);

            vector<float> dist = norm2d_axis0(sub);
            vector<bool> mask_dist = mask1d(dist, meta.decoder_min_distance / meta.stride);
            nine_filtered = filter_vector2D(nine_filtered, mask_dist);
        }

        if(this->meta.decoder_max_distance) {
            const Vector2D nine1 {nine_filtered[1], nine_filtered[2]};
            const Vector2D nine2 {nine_filtered[3], nine_filtered[4]};
            Vector2D sub = subtract2d(nine1, nine2);

            vector<float> dist = norm2d_axis0(sub);
            vector<bool> mask_dist = mask1d_lt(dist, this->meta.decoder_max_distance / this->meta.stride);
            nine_filtered = filter_vector2D(nine_filtered, mask_dist);
        }

        int n = nine_filtered.size();
        int m = nine_filtered[0].size();
        for(int j = 1; j < n; j++) {
            for(int k = 0; k < m; k++) {
                nine_filtered[j][k] *= this->meta.stride;
            }
        }

        vector<unsigned> nine_b_is {0, 3, 4, 1, 2, 6, 5, 8, 7};
        Vector2D nine_b;
        for(int j = 0; j < nine_b_is.size(); j++) {
            unsigned index = nine_b_is[j];
            nine_b.push_back(nine_filtered[index]);
        }
        Vector2D nine_b_rescored = this->rescore(nine_b, this->meta.head_metas.skeleton[i].first - 1);
        this->backward.push_back(nine_b_rescored);

        Vector2D nine_f = nine_filtered;
        Vector2D nine_f_rescored = this->rescore(nine_f, this->meta.head_metas.skeleton[i].second - 1);
        this->forward.push_back(nine_f_rescored);
    }
}

CafScored CafScored::fill(const Fields &all_fields) {
    this->fill_single(all_fields);

    return *this;
}